keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+         name: mysql-persistent-storage
+       spec:
+         accessModes: [ "ReadWriteOnce" ]
+         resources:
+           requests:
+             storage: 250Gi
+--- 
+apiVersion: v1
+Kind: Application
+metadata:
+  name: wordpress
+spec:
+  type: wordpress
+  version: 1.0
+  url: 
+  selector:
+    matchLables:
+      app: wordpress
+  components:
+    - wordpress-hsvc:
+        version: v1
+        kind: Service
+    - wordpress-svc:
+        version: v1
+        kind: Service
+    - wordpress-webserver:
+        group: apps
+        version: v1
+        kind: StatefulSet
+   dependencies:
+    - mysql
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: wordpress-svc
+  labels:
+    app: wordpress
+    component: wordpress-svc
+  annotations:
+      kubernetes.io/application: wordpress
+spec:
+  ports:
+    - port: 80
+  selector:
+    app: wordpress
+    component: wordpress-svc
+  type: LoadBalancer
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: wordpress-hsvc
+  labels:
+    app: wordpress
+    component: wordpress-hsvc
+  annotations:
+      kubernetes.io/application: wordpress
+spec:
+  ports:
+    - port: 3306
+  selector:
+    app: wordpress
+    component: wordpress-hsvc
+  clusterIP: None
+---
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: wordpress-webserver
+  labels:
+    app: wordpress
+    component: wordpress-webserver
+  annotations:
+      kubernetes.io/application: wordpress
+spec:
+  replicas: 1
+  service: wordpress-hsvc
+  selector:
+    matchLabels:
+      app: wordpress
+      component: wordpress-webserver
+  template:
+    metadata:
+      labels:
+        app: wordpress
+        tier: frontend
+    spec:
+      containers:
+      - image: wordpress:4.8-apache
+        name: wordpress
+        env:
+        - name: WORDPRESS_DB_HOST
+          value: wordpress-mysql
+        - name: WORDPRESS_DB_PASSWORD
+          valueFrom:
+            secretKeyRef:
+              name: mysql-pass
+              key: password
+        ports:
+        - containerPort: 80
+          name: wordpress
+        volumeMounts:
+        - name: wordpress-persistent-storage
+          mountPath: /var/www/html
+  volumeClaimTemplates:
+       - metadata:
+           name: mysql-persistent-storage
+         spec:
+           accessModes: [ "ReadWriteOnce" ]
+           resources:
+             requests:
+               storage: 250Gi
+```
+
+Above we generate two Applications, one for the MySQL RDBMs and another for the WordPress web server. Each of these 
+Applications can be managed independently. For instance, if the WordPress web server is deleted, the components of the 
+MySQL RDBMS will not be affected. However, the dependency between the two workloads is made explicit in the 
+`Spec.Dependencies` of the `wordpress` Application. Tools and UIs that examine the `wordpress` application are made 
+aware of its dependency on the `mysql` Application, and they can surface the dependency to users.
```

Comment:
This would imply that you have in mind a "weak dependency" model, where the application expects the dependency to be satisfied, but does not actively manage that dependency.

In this model, how do we...

- validate that a dependency is at the correct "version" that is required by this app
- handle the case where a dependency is upgraded to a version operationally incompatible with this application? (or any of the vast number of scenarios where a dependency can be altered in a way that makes it incompatible with the expectations of this app, like a change to labels that are being used as selectors)
- handle the case where a dependency is deleted after this app is installed?
- handle the case where an upgrade to the application should result in an upgrade/restart/etc to the dependency
- handle the case where the dependency exists, but the RBACs do not allow access

Given these, it's hard to understand what exactly is gained by declaring an explicit weak dependency at this level. Merely saying "I declare a dependency on an application named `mysql`" doesn't seem to me to add much value.

Now there's the case of strong dependencies. Say I want to require that a new MySQL is required for every Wordpress I install. The problem with the Application structure and this approach is that the top-level manifest doesn't allow for metadata on this -- there's only one license, one about url, one dashboard, etc. So while the operational part of strong dependencies are accommodated, they're sort of "hidden" dependencies, in the sense that they are not expressed as dependencies, but shunted into the main application. Given that some of these dependencies will have a sophisticated set of manifests that _could_ be maintained separately, this means that the complexity of the application manifests is somewhat compounded, and the tooling that manages the application will necessarily need to be more complex.

It might be useful to contrast this with Helm. Helm provides strong dependencies (in the form of subcharts), each of which has its own identity, metadata, configuration, etc. A Helm release manages both the app and all of its (strong) dependencies. But Helm does not define weak dependencies as an explicit thing, as there is no useful abstraction upon which we could say "that dependency is satisfied" for the duration of the application lifecycle. Instead, users are encouraged to use things like Kubernetes Service and ConfigMap objects to describe application-layer external dependencies. We've looked at various ways to accommodate weak dependencies, but the base problem is that the release itself underdetermines what is required to satisfy weak dependencies. E.g. the requirement to have a MySQL-compatbile DB can be equally satisfied by multiple different databases (mysql, percona, maria...) as well as by hosted solution (RDB...), and by service brokers.